ISPN-12108 configuring mediatypes client interop (#8554)
infinispan/infinispan: documentation/src/main/asciidoc/topics/proc_configuring_textbased_encoding.adoc

[id='using_text_based_storage-{context}']
= Using Text-Based Formats for Data
You can configure {brandname} to encode data in text-based formats such as
`text/plain`, `application/json`, or `application/xml`.

.Procedure

. Specify a text-based storage format as the MediaType for keys and values.
. Optionally specify a character set such as `UTF-8`.
+
The following example configures {brandname} to store entries with the `text/plain; charset=UTF-8` format:
+
[source,xml,options="nowrap",subs=attributes+]
----
include::config_examples/encoding_text_plain.xml[]
----
+
. Configure your clients.

> Hot Rod clients should use the `org.infinispan.commons.marshall.StringMarshaller` marshaller to handle data in text-based formats.

Author: @gustavonalle is this still correct for Hot Rod clients or should it change to ProtoStream?

Author: @ryanemerson same question here, please. for text-based formats should we change StringMarshaller to ProtoStream?

Reviewer: I'm not sure what use-case string based is trying to appeal to. If it's simply getting something working as simple as possible, then the `StringMarshaller` is the way to go. Protostream is more flexible, but for text based it has to be JSON and the user has to provide/generate protostream marshallers and a schema, so it's a lot more involved.

Author: hey @ryanemerson awesome. this helps clear it up a little for me. I've reworded to include some of this detail. IMO we should include ProtoStream in here too for consistency. also, when you say "Protostream is more flexible" - I take this to mean that you're pretty much restricted to string-based if you use StringMarshaller. that might not make much sense for Hot Rod clients. so if you want text + binary use ProtoStream. but, in this case, text means you must use the application/json mediatype.

does that sound right?

Reviewer: > I take this to mean that you're pretty much restricted to string-based if you use StringMarshaller. that might not make much sense for Hot Rod clients.

Exactly.

> so if you want text + binary use ProtoStream. but, in this case, text means you must use the application/json mediatype.

I was actually wrong on this. You can still specify [`text/plain`, `application/json`, 'application/x-java-object', 'application/octet-stream']  media types and transcoding will handle it, as it converts the value to a Java object if possible and then marshalls this using Protostream. XML is not supported though.

[id='using_text_based_storage-{context}']
= Using Text-Based Formats for Data
You can configure {brandname} to encode data in text-based formats such as
`text/plain`, `application/json`, or `application/xml`.

.Procedure

. Specify a text-based storage format as the MediaType for keys and values.
. Optionally specify a character set such as `UTF-8`.
+
The following example configures {brandname} to store entries with the `text/plain; charset=UTF-8` format:
+
[source,xml,options="nowrap",subs=attributes+]
----
include::config_examples/encoding_text_plain.xml[]
----
+
. Configure your clients.

Hot Rod clients can use the

* `org.infinispan.commons.marshall.StringMarshaller` handles plain text, JSON, XML, or any other text-based format.
* ProtoStream can handle JSON only but offers more flexibility for storing data in binary format. However, you must create Serialization Contexts and provide Protobuf schemas (`.proto` files) to use ProtoStream.

> 

Reviewer 2: Some nuance here: for text based format, protostream can also be used without registering proto files :)

[id='using_text_based_storage-{context}']
= Using Text-Based Formats for Data

> You can configure {brandname} to encode data in text-based formats such as

Reviewer 2: Better say "store data" instead to "encode data"

`text/plain`, `application/json`, or `application/xml`.

.Procedure

. Specify a text-based storage format as the MediaType for keys and values.
. Optionally specify a character set such as `UTF-8`.
+
The following example configures {brandname} to store entries with the `text/plain; charset=UTF-8` format:
+
[source,xml,options="nowrap",subs=attributes+]
----
include::config_examples/encoding_text_plain.xml[]
----
+
. Configure your clients.

Hot Rod clients can use `org.infinispan.commons.marshall.StringMarshaller`
to handle plain text, JSON, XML, or any other text-based format.

> If you want to use text-based formats and binary data with Hot Rod clients, you
> should use ProtoStream marshalling. ProtoStream can handle plain text and JSON 
> without the need to create Serialization Contexts and register Protobuf
> schemas. However, ProtoStream does not support XML.

Reviewer 2: ```suggestion
Text-based formats can also be used with the Protostream marshaller in the Hot Rod client
without the need to create Serialization Contexts and register Protobuf
schemas, since it can handle ```String``` and ```byte[]``` types natively
```

> ProtoStream offers more flexibility for storing data in binary format. However, you must  and provide Protobuf schemas (`.proto` files) to use ProtoStream.

Reviewer 2: I sense some duality between "text" and "binary" throughout the docs. 

Let me clarify: "text" is also binary for all it matters. ```text/plain```, ```application/json```, ```application/xml```, are all stored as ```byte[]``` inside the cache. The media type is just a description of what is inside that ```byte[]```.

Reviewer 2: Also, I think we haven't been camel casing ```Protostream```

Author: hmm. I've used camel case for ProtoStream elsewhere in docs. it also appears that way at: https://javadoc.io/doc/org.infinispan.protostream/protostream/latest/index.html

Reviewer 2: Sorry, I was wrong. Even ProtoStream itself uses it :smile:
